File: Leetcode/226.Invert_Binary_Tree/226.cpp
/* ************************************************************************
> File Name:     226.cpp
> Author:        Jack Yang
> Created Time:  2023年05月04日 (週四) 20時36分25秒
> Description:   Invert Binary Tree C++ iteration 
 ************************************************************************/
#include<iostream>
#include<stack>
#include<queue>
using namespace std;

struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };
typedef TreeNode TreeNode;
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        stack<TreeNode *>st;
        if(root)st.push(root);
        else if(!root)return nullptr;
        TreeNode *cur=root;
        while(!st.empty()){
            TreeNode *node=st.top();
            st.pop();
            swap(node->right,node->left);
            if(node->right)st.push(node->right);
            if(node->left)st.push(node->left);
        }
        return root;
    }
};

int main(){
    Solution sol;
    TreeNode *root=new TreeNode(4);
    TreeNode *ptr=root;
    ptr->left=new TreeNode(2);
    ptr->right=new TreeNode(7);
    ptr=ptr->left;
    ptr->left=new TreeNode(1);
    ptr->right=new TreeNode(3);
    ptr=root;
    ptr->right=new TreeNode(7);
    ptr=ptr->right;
    ptr->left=new TreeNode(6);
    ptr->right=new TreeNode(9);
    root=sol.invertTree(root);
    queue<TreeNode *>que;
    que.push(root);
    while(!que.empty()){
        TreeNode *node;
        node=que.front();
        que.pop();
        cout << node->val << endl;
        if(node->left)que.push(node->left);
        if(node->right)que.push(node->right);
    }
    return 0;
}

File: Leetcode/226. Invert Binary Tree/JavaScript/226.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

var invertTree = function (root) {
  let nodeArr = [];
  if (root) {
    nodeArr.push(root);
  }
  while (nodeArr.length > 0) {
    // 更新每層node數
    let size = nodeArr.length;
    // 遍歷單層node
    for (let i = 0; i < size; i++) {
      let node = nodeArr.shift();
      let left = node.left;
      let right = node.right;
      // 因提議左右子樹相反，所以push array時要相反
      if (right) {
        nodeArr.push(node.right);
      }

      if (left) {
        nodeArr.push(node.left);
      }
      // 翻轉子結點
      node.right = left;
      node.left = right;
    }
  }
  return root;
};

let root = new TreeNode(4, new TreeNode(2), new TreeNode(7));
root.left.left = new TreeNode(1);
root.left.right = new TreeNode(3);
root.right.left = new TreeNode(6);
root.right.right = new TreeNode(9);
invertTree(root);
console.log(root);
